src/Umbraco.Core/PropertyEditors/DataValueReferenceFactoryCollection.cs

```diff
@@ -32,15 +32,23 @@ public ISet<UmbracoEntityReference> GetAllReferences(IPropertyCollection propert
     {
         var references = new HashSet<UmbracoEntityReference>();
 
-        foreach (IProperty property in properties)
+        // Group by property editor alias to avoid duplicate lookups and optimize value parsing
+        foreach (var propertyValuesByPropertyEditorAlias in properties.GroupBy(x => x.PropertyType.PropertyEditorAlias, x => x.Values))
         {
-            if (!propertyEditors.TryGet(property.PropertyType.PropertyEditorAlias, out IDataEditor? dataEditor))
+            if (!propertyEditors.TryGet(propertyValuesByPropertyEditorAlias.Key, out IDataEditor? dataEditor))
             {
                 continue;
             }
 
-            // Only use edited value for now
-            references.UnionWith(GetReferences(dataEditor, property.Values.Select(x => x.EditedValue)));
+            // Use distinct values to avoid duplicate parsing of the same value
+            var values = new HashSet<object?>(properties.Count);
+            foreach (IPropertyValue propertyValue in propertyValuesByPropertyEditorAlias.SelectMany(x => x))
+            {
+                values.Add(propertyValue.EditedValue);
+                values.Add(propertyValue.PublishedValue);
+            }
```

Comment:
I've added grouping by property editor alias to avoid duplicate lookups (and creating the value editors) and optimized value parsing by only passing distinct/unique values to `GetReferences()`, because the same value will always return the same references.

Reply:
And as mentioned in https://github.com/umbraco/Umbraco-CMS/pull/15593#discussion_r1457292117, this also parses the published value, so references that are removed from the current draft (but still published) will still be tracked. This ensures you still get a warning when trying to delete e.g. a media item that is still used on a published page...